src/coronawhy_vt/utils/local_processing.py

```diff
@@ -0,0 +1,375 @@
+                    newname = name.replace(" ", "_").lower()
+                    ngram_map[name.lower()] = newname
+                    names.append(newname)
+            else:
+                names.append(name.lower())
+    names = np.unique(names)
+    return names, ngram_map
+
+
+# TODO Pull the next 6 functions into a class?
+# These functions determine what blocks are pulled from the paper for matching
+def title_blocks(paper: dict) -> list:
+    """Retrieve title block of a paper in json form.
+
+    :param paper (json): The paper to be parsed
+
+    :return: Lines of text in the paper title
+    """
+    return [{'text': paper['metadata']['title']}]
+
+
+def abstract_blocks(paper: dict) -> list:
+    """Retrieve abstract block of a paper in json form.
+
+    :param paper: The paper to be parsed
+
+    :return: Lines of text in the paper abstract
+    """
+    return paper['abstract']
+
+
+def body_blocks(paper: dict) -> list:
+    """Retrieve body block of a paper in json form.
+
+    :param paper: The paper to be parsed
+
+    :return: Lines of text in the paper body
+    """
+    return paper['body_text']
+
+
+# TODO the reference nlp is an external dependence, either add into function or make it a class
+nlp = spacy.load('en_core_web_lg')
+
+
+def pull_mentions_lemmatized(
+        paths: list,
+        block_selector: Callable[[dict], list],
+        sec_name: str,
+        words: list,
+        replace_dict: dict = None) -> dict:
+    """Aggregate the positional features of the lemmatized text corpus.
+
+    :param paths: pathlib.Path objects corresponding to directories containing
+        corpus to aggregate features from.
+    :param block_selector: The function to retrieve the relevant block from text
+    :param sec_name: Corresponds with block_selector to select the relevant
+        block of text. Options are 'title', 'abstract', 'body'
+        TODO seems like this should be pulled into the function as it depends
+        entirely on the block_selector argument
+    :param words: List of lemmatized words
+    :param replace_dict: Takes a dict of {token: lemmatized underscore token}
+        and replaces occurences of token in text with its value
+
+    :return: {'identifier feature': list of occurrences}
+    """
+    positions = []
+    found_words = []
+    section = []
+    block_id = []
+    block_text = []
+    paper_id = []
+
+    tokenized_words = []
+    for w in words:
+        tokenized_words.append(nlp(w.lower())[0].lemma_)
+    for path in paths:
+        files = os.listdir(path)
+        for p in files:
+
+            readfile = open(path / p, 'r')
+            paper = json.load(readfile)
+            blocks = block_selector(paper)
+
+            for b in range(0, len(blocks)):
+                text = blocks[b]['text'].lower()
+                if(replace_dict is not None):
+                    text = run_replace(text, replace_dict)
+                text = nlp(text)
+                for t in text:
```

Comment:
since you're going two levels deep in independent iterators, it's better to use `itertools`

```python
import itertools as itt
for t, w in itt.prodict(text, tokenized_words):
    ...
```

don't worry - the iteration will go in exactly the same order